<a href="https://colab.research.google.com/github/benitoredin0/BootCamp-2/blob/main/Pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<p style="text-align: center;"> <font size="6">💘🐼🙏🐼 PANDAS 🐼🙏🐼💘 </font> </p>


![](https://img-9gag-fun.9cache.com/photo/a61Q85R_460s.jpg)



https://pandas.pydata.org/docs/user_guide/index.html

In [ ]:
#pip install pandas

🐼 Para tener acceso a Pandas hay que importar el módulo pandas. Por convención, se importa con el nombre **pd**

In [ ]:
import pandas as pd

🐼 Crear un dataframe

In [18]:
precio = [2000, 12300, 43245, 325423, 200]
ncomuna = [2, 5, 3, 4, 5]
ambientes = [2, 1, 4, 5, 6]

In [19]:
tabla = {
    "precio":precio,
    "ncomuna":ncomuna,
    "ambientes":ambientes
        }

In [20]:
df = pd.DataFrame(tabla)
print(df)

   precio  ncomuna  ambientes
0    2000        2          2
1   12300        5          1
2   43245        3          4
3  325423        4          5
4     200        5          6


🐼 Con la instruccion anterior creamos un objeto llamado ``DataFrame`` en la variable ``df``. Este es el **núcleo** de Pandas y vamos a explorar muchas de las cosas que se puede hacer a lo largo de las clases.

🐼 Entonces, un DataFrame es una tabla de datos, tiene columnas y filas. En este caso particular, cada fila o registro del dataframe pertenece a una propiedad, teniendo como informacion:

- Su precio
- El numero de comuna a la que pertenece
- El numero de ambientes que posee.

🐼 Podemos ver que hay algunas partes clave de la salida:

- En el lado izquierdo en negrita está el índice. Estos pueden considerarse como números de fila y como todo indice identifica a su fila de forma unica (no se repiten). En este caso simplemente el indice es un incremental.
- En la parte superior están los nombres de las columnas. Cuando queramos referirnos a una columna en particular en nuestro DataFrame, usaremos estos nombres.
- A continuación, los datos reales se organizan en el centro de la tabla. En su mayoría, estos son datos que nos importan.

🐼 Podemos imprimir la cantidad de filas y columnas que tiene el DataFrame

In [21]:
print(df.precio)

0      2000
1     12300
2     43245
3    325423
4       200
Name: precio, dtype: int64


Ponele voluntad...

🐼 Podemos imprimir el nombre de las columnas

In [9]:
df.columns

Index(['precio', 'ncomuna', 'ambientes'], dtype='object')

🐼 Ejercicio: poner en mayusculas el nombre de todas las columnas

In [22]:
df.columns = df.columns.str.upper()
df.columns

Index(['PRECIO', 'NCOMUNA', 'AMBIENTES'], dtype='object')

Y si tenemos 300 columnas? se utiliza FOR:

🐼 Ejercicio: Modificar el indice por esta lista los siguientes elementos: [10, 9, 8, 7, "pepe"]

In [23]:
nuevo_indice =  [10, 9, 8, 7, "pepe"]
df = df.set_index(pd.Index(nuevo_indice))

print(df)

      PRECIO  NCOMUNA  AMBIENTES
10      2000        2          2
9      12300        5          1
8      43245        3          4
7     325423        4          5
pepe     200        5          6


🐼 Resetear el indice para que vuelva a estar entre 0 y n

In [24]:
df = df.reset_index(drop=True)
print(df)

   PRECIO  NCOMUNA  AMBIENTES
0    2000        2          2
1   12300        5          1
2   43245        3          4
3  325423        4          5
4     200        5          6


🐼 Importar un csv que tengamos en nuestro local o en una direccion web

In [25]:
departamentos = pd.read_csv('https://raw.githubusercontent.com/ConroSatchmo/datasets/main/departamentos.csv')

In [26]:
departamentos

M2  AMBIENTES  ANTIGUEDAD  BAÑOS    LATITUD   LONGITUD     COMUNA  \
0      81          3           4      1 -34.581078 -58.449433  COMUNA 13   
1      69          3          20      1 -34.623129 -58.439338  COMUNA 06   
2      75          3          20      1 -34.604972 -58.421278  COMUNA 05   
3      42          2          40      1 -34.604725 -58.399524  COMUNA 03   
4      90          3           1      1 -34.623390 -58.504401  COMUNA 10   
...   ...        ...         ...    ...        ...        ...        ...   
2227   88          4          20      2 -34.550152 -58.470136  COMUNA 13   
2228   35          1          35      1 -34.602256 -58.429315  COMUNA 05   
2229  117          3          10      2 -34.620314 -58.365451  COMUNA 01   
2230   69          4          40      1 -34.628533 -58.427748  COMUNA 07   
2231   73          4          50      2 -34.596659 -58.378396  COMUNA 01   

      DOLARES  
0      225000  
1      140000  
2      154000  
3       75000  
4      149900  
...       ...  
2227   165000  
2228    79900  
2229   520000  
2230   132000  
2231   179000  

[2232 rows x 8 columns]

🐼 Supongamos que solo me interesa la columna dolares

In [27]:
departamentos.DOLARES

0       225000
1       140000
2       154000
3        75000
4       149900
         ...  
2227    165000
2228     79900
2229    520000
2230    132000
2231    179000
Name: DOLARES, Length: 2232, dtype: int64

Un objeto ``Series`` es un objeto unidimensional, **similar a una columna en una tabla**.

Un objeto de tipo Series tiene dos componentes principales:

- Un indice, que contiene valores unicos y se usa para acceder a los valores. Por defecto, cada elemento recibirá una etiqueta que va de 0 a N-1, donde N es la longitud/tamaño de la colección.
- Los valores.

**Un conjunto de Series forman un DataFrame**

Como nos sentimos mas comodos con los DataFrame que con las Series, podemos pasar le obejeto ``Series`` a ``DataFrame``

##### 🐼 Imprimir los primeros 5 registros/filas

In [31]:
filtro_5 = departamentos.iloc[:5]
filtro_5

M2  AMBIENTES  ANTIGUEDAD  BAÑOS    LATITUD   LONGITUD     COMUNA  DOLARES
0  81          3           4      1 -34.581078 -58.449433  COMUNA 13   225000
1  69          3          20      1 -34.623129 -58.439338  COMUNA 06   140000
2  75          3          20      1 -34.604972 -58.421278  COMUNA 05   154000
3  42          2          40      1 -34.604725 -58.399524  COMUNA 03    75000
4  90          3           1      1 -34.623390 -58.504401  COMUNA 10   149900

In [34]:
filtro_5a10 = departamentos.iloc[5:11]
filtro_5a10

M2  AMBIENTES  ANTIGUEDAD  BAÑOS    LATITUD   LONGITUD     COMUNA  \
5   120          3           2      2 -34.577774 -58.424300  COMUNA 14   
6    35          1           5      1 -34.548911 -58.469872  COMUNA 12   
7    82          3          70      2 -34.637083 -58.405029  COMUNA 04   
8    38          1           8      1 -34.580970 -58.440528  COMUNA 14   
9    17          2          40      1 -34.587077 -58.401069  COMUNA 02   
10   45          2          10      1 -34.592452 -58.401854  COMUNA 02   

    DOLARES  
5    630000  
6     92000  
7    135000  
8    110000  
9     45000  
10   145000

🐼 Imprimir en pantalla la cantidad de M2 que tiene el primer registro

In [36]:
departamentos.M2.iloc[0]

81

Le ponemos un poco de onda:

🐼 Imprimir los ultimos 5 registros

In [37]:
filtro_ult5 = departamentos.iloc[-5:]
filtro_ult5

M2  AMBIENTES  ANTIGUEDAD  BAÑOS    LATITUD   LONGITUD     COMUNA  \
2227   88          4          20      2 -34.550152 -58.470136  COMUNA 13   
2228   35          1          35      1 -34.602256 -58.429315  COMUNA 05   
2229  117          3          10      2 -34.620314 -58.365451  COMUNA 01   
2230   69          4          40      1 -34.628533 -58.427748  COMUNA 07   
2231   73          4          50      2 -34.596659 -58.378396  COMUNA 01   

      DOLARES  
2227   165000  
2228    79900  
2229   520000  
2230   132000  
2231   179000

🐼 Crear una nueva columna llamada PAIS y rellenarla con "Argentina"

In [38]:
departamentos['PAIS'] = 'Argentina'
departamentos

M2  AMBIENTES  ANTIGUEDAD  BAÑOS    LATITUD   LONGITUD     COMUNA  \
0      81          3           4      1 -34.581078 -58.449433  COMUNA 13   
1      69          3          20      1 -34.623129 -58.439338  COMUNA 06   
2      75          3          20      1 -34.604972 -58.421278  COMUNA 05   
3      42          2          40      1 -34.604725 -58.399524  COMUNA 03   
4      90          3           1      1 -34.623390 -58.504401  COMUNA 10   
...   ...        ...         ...    ...        ...        ...        ...   
2227   88          4          20      2 -34.550152 -58.470136  COMUNA 13   
2228   35          1          35      1 -34.602256 -58.429315  COMUNA 05   
2229  117          3          10      2 -34.620314 -58.365451  COMUNA 01   
2230   69          4          40      1 -34.628533 -58.427748  COMUNA 07   
2231   73          4          50      2 -34.596659 -58.378396  COMUNA 01   

      DOLARES       PAIS  
0      225000  Argentina  
1      140000  Argentina  
2      154000  Argentina  
3       75000  Argentina  
4      149900  Argentina  
...       ...        ...  
2227   165000  Argentina  
2228    79900  Argentina  
2229   520000  Argentina  
2230   132000  Argentina  
2231   179000  Argentina  

[2232 rows x 9 columns]

🐼 Utiliza .info() para tener informacion muy importante y util del dataframe

In [39]:
departamentos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2232 entries, 0 to 2231
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   M2          2232 non-null   int64  
 1   AMBIENTES   2232 non-null   int64  
 2   ANTIGUEDAD  2232 non-null   int64  
 3   BAÑOS       2232 non-null   int64  
 4   LATITUD     2232 non-null   float64
 5   LONGITUD    2232 non-null   float64
 6   COMUNA      2232 non-null   object 
 7   DOLARES     2232 non-null   int64  
 8   PAIS        2232 non-null   object 
dtypes: float64(2), int64(5), object(2)
memory usage: 157.1+ KB


🐼 Generar una columna que se llame PESOS cuyo valor sea DOLARES * 200

In [40]:
departamentos['PESOS'] = departamentos['DOLARES'] * 200
departamentos

M2  AMBIENTES  ANTIGUEDAD  BAÑOS    LATITUD   LONGITUD     COMUNA  \
0      81          3           4      1 -34.581078 -58.449433  COMUNA 13   
1      69          3          20      1 -34.623129 -58.439338  COMUNA 06   
2      75          3          20      1 -34.604972 -58.421278  COMUNA 05   
3      42          2          40      1 -34.604725 -58.399524  COMUNA 03   
4      90          3           1      1 -34.623390 -58.504401  COMUNA 10   
...   ...        ...         ...    ...        ...        ...        ...   
2227   88          4          20      2 -34.550152 -58.470136  COMUNA 13   
2228   35          1          35      1 -34.602256 -58.429315  COMUNA 05   
2229  117          3          10      2 -34.620314 -58.365451  COMUNA 01   
2230   69          4          40      1 -34.628533 -58.427748  COMUNA 07   
2231   73          4          50      2 -34.596659 -58.378396  COMUNA 01   

      DOLARES       PAIS      PESOS  
0      225000  Argentina   45000000  
1      140000  Argentina   28000000  
2      154000  Argentina   30800000  
3       75000  Argentina   15000000  
4      149900  Argentina   29980000  
...       ...        ...        ...  
2227   165000  Argentina   33000000  
2228    79900  Argentina   15980000  
2229   520000  Argentina  104000000  
2230   132000  Argentina   26400000  
2231   179000  Argentina   35800000  

[2232 rows x 10 columns]

🐼 Imprimi los datos estadisticos del dataframe

In [42]:
desc_est = departamentos.describe()
print(desc_est)

                M2    AMBIENTES   ANTIGUEDAD        BAÑOS      LATITUD  \
count  2232.000000  2232.000000  2232.000000  2232.000000  2232.000000   
mean     79.396057     2.807796    24.633513     1.430108   -34.592671   
std      58.977534     1.109541    18.408973     0.720170     0.022613   
min      15.000000     1.000000     1.000000     1.000000   -34.684919   
25%      44.000000     2.000000     9.000000     1.000000   -34.606372   
50%      64.000000     3.000000    20.000000     1.000000   -34.592954   
75%      90.000000     3.000000    40.000000     2.000000   -34.577447   
max     730.000000    10.000000   110.000000     6.000000   -34.537308   

          LONGITUD       DOLARES         PESOS  
count  2232.000000  2.232000e+03  2.232000e+03  
mean    -58.428680  2.060140e+05  4.120280e+07  
std       0.034102  1.863683e+05  3.727366e+07  
min     -58.527476  7.428000e+03  1.485600e+06  
25%     -58.453970  9.837500e+04  1.967500e+07  
50%     -58.424798  1.460000e+05  2.920

Podemos ver que la **media** de las posiciones esta en [-34.592671,	-58.428680], es decir las propiedades de mi DataFrame "orbitan" sobre este punto.

El **desvio estandar** es de [0.022613,	0.034102]

*El desvio estandar es una medida que se utiliza para cuantificar la dispersión de un conjunto de datos numéricos.*

![](https://upload.wikimedia.org/wikipedia/commons/thumb/8/8c/Standard_deviation_diagram.svg/450px-Standard_deviation_diagram.svg.png)

Vemos que la mayoria de registros pertenecen a CABA:

Verificar posibles valores nulos

🐼 Imprimir el valor que se encuentra en el indice 5 y en el indice 1 y columna M2

In [45]:
valor_5=departamentos.M2.iloc[5]
valor_1=departamentos.M2.iloc[1]

print(valor_5, valor_1)

120 69


🐼 Obtener los valores unicos que integran la columna ``COMUNA``

In [46]:
valores_unicos = departamentos['COMUNA'].unique()
print(valores_unicos
      )

['COMUNA 13' 'COMUNA 06' 'COMUNA 05' 'COMUNA 03' 'COMUNA 10' 'COMUNA 14'
 'COMUNA 12' 'COMUNA 04' 'COMUNA 02' 'COMUNA 01' 'COMUNA 07' 'COMUNA 11'
 'COMUNA 15' 'COMUNA 08' 'COMUNA 09']


🐼 Imprimir las primeras 3 columnas y las primeras 4 filas

In [47]:
filtro = departamentos.iloc[:4,:3]
print(filtro)

   M2  AMBIENTES  ANTIGUEDAD
0  81          3           4
1  69          3          20
2  75          3          20
3  42          2          40


🐼 Imprimir el valor de la columna 5 y fila 2

In [54]:
filtro1 = departamentos.iat[1,4]
print(filtro1)

-34.62312877177625


🐼 Seleccionar todas las filas de las siguientes columnas: DOLARES y COMUNA

In [57]:
print(departamentos[['DOLARES','COMUNA']])

      DOLARES     COMUNA
0      225000  COMUNA 13
1      140000  COMUNA 06
2      154000  COMUNA 05
3       75000  COMUNA 03
4      149900  COMUNA 10
...       ...        ...
2227   165000  COMUNA 13
2228    79900  COMUNA 05
2229   520000  COMUNA 01
2230   132000  COMUNA 07
2231   179000  COMUNA 01

[2232 rows x 2 columns]


🐼 **Filtrado**: Seleccionar las filas que sean de la comuna 13

In [59]:
filtro_comuna13 = departamentos[departamentos['COMUNA'] == 'COMUNA 13']
print(filtro_comuna13)

      M2  AMBIENTES  ANTIGUEDAD  BAÑOS    LATITUD   LONGITUD     COMUNA  \
0     81          3           4      1 -34.581078 -58.449433  COMUNA 13   
21    34          1           5      1 -34.564887 -58.461323  COMUNA 13   
26    42          2          10      1 -34.555101 -58.461095  COMUNA 13   
27    33          1          20      1 -34.537308 -58.467448  COMUNA 13   
40    68          3          20      1 -34.561799 -58.444212  COMUNA 13   
...   ..        ...         ...    ...        ...        ...        ...   
2183  88          4          20      2 -34.550152 -58.470136  COMUNA 13   
2186  45          2           2      1 -34.567289 -58.456200  COMUNA 13   
2197  45          2          15      1 -34.553523 -58.466288  COMUNA 13   
2203  64          3          40      1 -34.544475 -58.469458  COMUNA 13   
2227  88          4          20      2 -34.550152 -58.470136  COMUNA 13   

      DOLARES       PAIS     PESOS  
0      225000  Argentina  45000000  
21     109000  Argentina 

🐼 Selecccionar todas las filas de las columnas que empiezan con 'A'

In [63]:
colum_filtro_A = departamentos.filter(regex='^A')

filas_filtro_A = departamentos[colum_filtro_A.columns]

print(filas_filtro_A)

      AMBIENTES  ANTIGUEDAD
0             3           4
1             3          20
2             3          20
3             2          40
4             3           1
...         ...         ...
2227          4          20
2228          1          35
2229          3          10
2230          4          40
2231          4          50

[2232 rows x 2 columns]


      M2  AMBIENTES  ANTIGUEDAD  BAÑOS   LATITUD   LONGITUD     COMUNA  \
1530  15          2          20      1 -34.55782 -58.477648  COMUNA 12   

      DOLARES       PAIS    PESOS  
1530    35000  Argentina  7000000  


🐼 Seleccionar los departamentos que valgan menos de $50000 **o** esten en *COMUNA 02*

In [76]:
condicion = (departamentos['DOLARES'] < 50000) & (departamentos['COMUNA'] == 'COMUNA 12')
dptos = departamentos[condicion]
print(dptos)

      M2  AMBIENTES  ANTIGUEDAD  BAÑOS   LATITUD   LONGITUD     COMUNA  \
1530  15          2          20      1 -34.55782 -58.477648  COMUNA 12   

      DOLARES       PAIS    PESOS  
1530    35000  Argentina  7000000  


🐼 Seleccionar los departamentos que tengan mas de 3 baños **y** cuya antiguedad sea menor a 5 años

🐼 Seleccionar los departamentos que **no** pertenezcan a la COMUNA 13

🐼 Seleccionar los departamentos cuyo valor se encuentre entre 50000 y 60000

🐼 Obtener el departamento con menor precio que pertenezca a alguna de las siguientes comunas: "COMUNA 03", "COMUNA 07", "COMUNA 15"

🐼 La verdadera potencia de pandas se demuestra cuando **encadenamos** operaciones

🐼 Contar la cantidad de departamentos que hay en cada comuna.

🐼 Contar por ambientes los departamentos de la comuna 14

🐼Metodo para estadistica descriptiva

- sum()
- mean()
- max()
- min()

🐼 Un poco de visualización

📈 Grafico de puntos: nos demuestra relaciones entre las variables

📈 Grafico de distribución: En lugar de ver cómo se relaciona una variable con otra, a veces queremos ver cómo se organizan los puntos de datos en una dimensión particular.